### PR TITLE
Switch default multiprocessing context to "spawn"

### LINKED ISSUE #3759: Default `multiprocessing` context is broken and should never be used
The default `multiprocessing` context (on POSIX systems) is fork-without-exec. As a result, all process state is mirrored in the subprocesses, with often unexpected results.

1. Threads are not propagated. If the parent process started a thread and code relies on it working this can break the subprocess. This is probably fairly rare, though I do maintain a library where this is a problem.
2. More seriously, module-level globals are preserved. Common module-level state includes things like logging. Subprocesses inherit the parent process' logging, often with unexpected results. I have seen child processes trying to rotate the parent's log file, for example.
3. Even if dropping threads doesn't directly impact the subprocess, it can indirectly cause deadlocks. In particular, if it acquires a lock and a fork happens immediately after that, the lock will be acquired in the subprocess and *never released*.

The following example (on Python 3.5) demonstrates problems 2 and 3. Notice how merely using the `logging` module in a thread can break subsequent processes created with `multiprocessing`:

```python
import logging
import threading
import multiprocessing
import sys
import time

logging.basicConfig(stream=sys.stdout, level=logging.INFO)
logging.info("HELLO FROM PARENT")

# Emulate a thread that is holding the logging lock when fork() happens. In real
# code this would be a race condition, I'm just making it fail consistently for
# demonstration purposes.
def logging_thread():
    logging._acquireLock()
    for i in range(1, 10):
        print("THREAD COUNTDOWN TO RELEASE LOCK", i)
        time.sleep(1)
    logging._releaseLock()
    print("THREAD IS GOING AWAY")

t = threading.Thread(target=logging_thread)
t.start()


def child_process():
    logging.info("CHILD PROCESS 1")
    # Do something that acquires lock:
    logging.Handler().set_name("hello")
    logging.info("CHILD PROCESS 2")

p = multiprocessing.Process(target=child_process)
p.start()
```

Here is the output:

```console
$ python3.6 mp-fork.py 
INFO:root:HELLO FROM PARENT
THREAD COUNTDOWN TO RELEASE LOCK 1
INFO:root:CHILD PROCESS 1
THREAD COUNTDOWN TO RELEASE LOCK 2
THREAD COUNTDOWN TO RELEASE LOCK 3
THREAD COUNTDOWN TO RELEASE LOCK 4
THREAD COUNTDOWN TO RELEASE LOCK 5
THREAD COUNTDOWN TO RELEASE LOCK 6
THREAD COUNTDOWN TO RELEASE LOCK 7
THREAD COUNTDOWN TO RELEASE LOCK 8
THREAD COUNTDOWN TO RELEASE LOCK 9
THREAD IS GOING AWAY
```

"CHILD PROCESS 2" is never printed because the child process is deadlocked, and the process never exits.

---

Proposed solution: change default multiprocessing context to `spawn`.

### PR DESCRIPTION
**Pull request by itamarst**
- [x] Tests added / passed
- [x] Passes `flake8 dask`

Fixes #3759.

Note that this may require some code to use `if __name__ == '__main__':` to start dask (https://docs.python.org/3/library/multiprocessing.html#the-spawn-and-forkserver-start-methods) in single module programs. `cloudpickle` probably makes this unnecessary, though?

### CODE REVIEW

**Comment by mrocklin** [comment]
I'm curious, how does this affect performance?  My guess is that it would pretty negatively affect things for modestly sized tasks.

**Comment by itamarst** [comment]
It only affects startup. Once the pool is started it doesn't make a difference. If pool startup is a bottleneck you probably shouldn't be using Dask, so that doesn't seem like a big deal to me.

**Comment by itamarst** [comment]
I wrote a little test program to check the above.

1. This change makes `if __name__ == __main__:` necessary for single module example to work, it seems.
2. The performance overhead is just small additional fixed cost at startup.

```python
import time
import sys
from dask.bag import from_sequence
from dask import compute, config

def expensive(x):
    time.sleep(1)
    return x * 2

if __name__ == '__main__':
    with config.set({"multiprocessing.context": sys.argv[1]}):
        start = time.time()
        bag = from_sequence(range(90)).map(expensive).sum()
        print(compute(bag, npartitions=8))
        print(f"Elapsed with method {sys.argv[1]}", time.time() - start)
```

```console
$ python3 benchmark.py fork
(8010,)
Elapsed with method fork 12.09099531173706
$ python3 benchmark.py spawn
(8010,)
Elapsed with method spawn 12.255864143371582
```

**Comment by mrocklin** [comment]
> The performance overhead is just small additional fixed cost at startup.

The tasks in your benchmark take 1s.  What if instead they take 10ms?  Is the cost still negligible?

I find myself not particularly excited about this change because I don't think that I've ever heard of a user running into a problem with the fork solution. I'm curious @itamarst have you run into issues with Dask's use of fork in practice, or are you promoting the use of good technology?  Are there common situations that you suspect occur frequently among dask users where this is an issue?

**Comment by mrocklin** [comment]
I'm not very well informed on this topic though.  It would be great if some other maintainers chimed in here.

**Comment by itamarst** [comment]
I have encountered—not in Dask, but in other uses of multiprocessing—random freezes that were caused by the default method, as well as messed up logging behavior. You won't always see them yourself because they're race conditions, but basically the experience is "my program is acting _wierd_ and I have no idea why." Most users will just give up, or just live with things freezing 1% of the time.

Basically, the default method is fundamentally broken. See the original issue, and a much longer writeup I did: https://codewithoutrules.com/2018/09/04/python-multiprocessing/

**Comment by mrocklin** [comment]
Dask.distributed workers have an intermittent failure like that as well in
Python 2.

However I've never run into anything in practice when using the
multiprocessing scheduler.  I wonder if that doesn't run into the same
issues somehow?  I would have expected it to have arisen in a bug report by
now.

**Comment by itamarst** [comment]
This change might break your code... but your code was previously broken some percentage of the time already...

Would it be acceptable if lack of `if __name__ == '__main__'` resulted in a clear error indicating the fix? Might be possible to detect that (maybe).

**Comment by jcrist** [comment]
> but your code was previously broken some percentage of the time already

This isn't necessarily true. Perhaps I only care about posix, and only use modules with fork-safe behavior. I suspect a non-negligible portion of dask users don't care about writing code that deploys on every OS, only that their code works on their machine.

I'm not 100% against this, but in 3.5 years we've never had a user run into this issue when using dask. I'm hesitant to make changes that will break user's code without clear evidence that the existing behavior is bad (and I'm not currently convinced that it is).

That said, I suspect at this point that most users use either the threaded scheduler or the distributed scheduler - the distributed scheduler runs fine locally, and is more efficient in many cases than the multiprocessing scheduler. Breaking code for a few to save many from having future issues may be worth it still.

**Comment by itamarst** [comment]
Another alternative: warn in visible way if users are using current default (permanently, or as prelude to eventual merge of this PR).

**Comment by mrocklin** [comment]
> warn in visible way if users are using current default

Building on what @jcrist has said above, the use of fork doesn't really seem to be bothering users today or stopping them from being productive.  We get bug reports about *lots* of parts of the system, but this topic has never come up.  I hesitate to teach people and push them away from a system that, pragmatically at least, seems to work ok today.  There is a very real cost to raising a warning to every user that I don't think is matched by any perceived benefits here.  I would want to revisit this if we found that many people were having difficulty due to using the defaults.

**Comment by itamarst** [comment]
OK.

**Comment by itamarst** [comment]
@jakirkham said he is in favor of making this change, this time around (https://github.com/dask/dask/issues/3759#issuecomment-619169780).

File: dask/multiprocessing.py [hunk 1]
```diff
@@ -143,14 +143,8 @@ def get_context():
         if config.get("multiprocessing.context", None) is not None:
             warn(_CONTEXT_UNSUPPORTED, UserWarning)
         return multiprocessing
```
Comment (jakirkham) [inline dask/multiprocessing.py:145]: On Windows the default (and only supported) context is `"spawn"`. Given this, can we drop this block?

**Comment by jakirkham** [comment]
To reiterate what I said there, we are already doing this in Distributed ( https://github.com/dask/distributed/pull/3461 ). This just does the same thing in Dask.

**Comment by TomAugspurger** [comment]
Thanks for reopening. IMO distributed using spawn by default tips the scales to also using `spawn` here, but we'll wait for feedback from others before merging..

File: docs/source/scheduler-overview.rst [hunk 2]
```diff
@@ -146,7 +146,7 @@ global pool set with ``dask.config.set``:
     ...     x.compute()
 
 The multiprocessing scheduler also supports `different contexts`_ ("spawn",
-"forkserver", "fork") which you can set with ``dask.config.set``:
+"forkserver", "fork") which you can set with ``dask.config.set``. The default context is "spawn", but you can set a different one:
```
Comment (jakirkham) [inline docs/source/scheduler-overview.rst:149]: May want to wrap this line as it's a bit long.

**Comment by TomAugspurger** [comment]
Seems to be some related test failures https://github.com/dask/dask/pull/4003/checks?check_run_id=616399950#step:6:197

```
____________________ test_custom_context_ignored_elsewhere ____________________

self = <multiprocessing.context.DefaultContext object at 0x000001BE1AF94518>
method = 'forkserver'

    def get_context(self, method=None):
        if method is None:
            return self
        try:
>           ctx = _concrete_contexts[method]
E           KeyError: 'forkserver'
```

and https://travis-ci.org/github/dask/dask/jobs/679167905#L1057

File: dask/multiprocessing.py [hunk 3]
```diff
@@ -138,19 +138,9 @@ def pack_exception(e, dumps):
...
-    # The default context on OSX was switched from "fork" to "spawn" in
-    # Python 3.8. We keep "fork" as the default here for backwards
-    # compatibility and to avoid raising a RuntimeError when using the
-    # multiprocessing scheduler in a Python script that is not in a
-    # if __name__ == "__main__" block
-    if context_name is None and sys.platform == "darwin":
-        context_name = "fork"
+    # fork context does fork()-without-exec(), which can lead to deadlocks,
+    # so default to "spawn".
+    context_name = config.get("multiprocessing.context", "spawn")
```
Comment (jakirkham) [inline dask/multiprocessing.py:143]: Ah ok. Seeing the test failure maybe we should just readd the previous Windows check after this line (tweaking it to check if it is not `"spawn"` instead).

**Comment by jakirkham** [comment]
Appears there is a linter failure. Could you please run `black` locally and push the changes?

**Comment by jcrist** [comment]
Yeah, I'm fine with this now. I think at this point we've successfully dissuaded most use of the multiprocessing scheduler in favor of distributed (which is significantly more efficient), so this change probably affects only a few users anyway.

**Comment by mrocklin** [comment]
I wouldn't wait for me on a decision here

**Comment by TomAugspurger** [comment]
Thanks!

**Comment by jakirkham** [comment]
Thanks @itamarst for the PR and everyone for the reviews! 😄
